### PR TITLE
ISSUE-462 # fix poll time out to intial consumer group join

### PR DESCRIPTION
**Pull request by dandalavinod**
# initialPollWaitingForConsumerGroupJoin time should be configurable

PR Branch
https://github.com/dandalavinod/zerocode/tree/issue462-polltimeout-kafkajoin

## Motivation and Context
Intial poll of the kafka consumer group is currently waiting for 500ms which in some cases, where there are network latency or broker throughput issues, fail with kafka consumer fail to join in time. 
The fix is to make the intial poll duration wait configurable to use the consumer.pollingTime field set in the properties file and if this property is not set then will default to current 500ms poll duration

## Checklist:

* [ x] Unit tests added
   updated existing unit tests to fit and behave per the changes added.

* [x] Integration tests added

* [x ] Test names are meaningful

* [ x] Feature manually tested
Tested manually against my environment where i encountered the problem and as i use the configurable setting of consumer.pollingTime=1000 ; the error i encounter very often is not reproducible now.

* [ x] Branch build passed

* [ x] No 'package.*' in the imports

* [ ] Relevant Wiki page updated with clear instruction for the end user
  * [ x] Not applicable. This was only a refactor change, no functional or behaviour changes were introduced

* [ ] Http test added to `http-testing` module(if applicable) ?
  * [ x] Not applicable. The changes did not affect HTTP automation flow

* [ ] Kafka test added to `kafka-testing` module(if applicable) ?
  * [x ] Not applicable. The changes did not affect Kafka automation flow

### CODE REVIEW

**Comment by authorjapps** [comment]
Solves #462

**Comment by authorjapps** [comment]
@ddingel, please have a look and review please when you get a chance ✌️

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (nirmalchandra) [inline core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:93]: Looks good @dandalavinod . Can you add a test(couple of tests) to cover this?

It should be fairly easy if you follow the earlier PR related to this (https://github.com/authorjapps/zerocode/pull/422).
Probably you can duplicate the earlier test and amend it to setup a new scenario with this config param. That's it.

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (ddingel) [inline core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:104]: I would suggest to remove this empty variation, as we always have the respective configuration.

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (dandalavinod) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:93]: @nirmalchandra thanks for the review. As you see, i have added/updated the current unit tests to support and verify fix. However, from integration level - i believe the current consume tests will verify my fix - since now for every consumer initial poll the code path i changed will be exercised and in turn be valdated ; as every consumer's consume call will tread through the intial poll join. Having said that - i added a integration test to verify case when the consumer.pollingTime is not provided in the properties file - please refer to my new commit related to these changes.

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (ddingel) [inline core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:87]: When you use ConsumerLocalConfigs with the derived configuration I assume you could use the `static getPollTime()` and skip defining a default, as there already exists one in KafkaConstants.

File: core/src/test/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelperTest.java
Comment (ddingel) [inline core/src/test/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelperTest.java:163]: Style thing: as this belongs to the given part of the test I would move this line beneath the comment.

**Review by ddingel — COMMENTED** [review]
@dandalavinod I like the idea! I added some suggestions which would result in a smaller change, please have a look :).

**Comment by ddingel** [comment]
One more thing, by using ConsumerConfigs we will using:
`Long DEFAULT_POLLING_TIME_MILLI_SEC = 100L;` as the default value instead of the hardcoded 500L.
Do we need to compensate by increasing the number of runs on the initial polling?

File: core/src/main/java/org/jsmart/zerocode/core/kafka/receive/KafkaReceiver.java
Comment (dandalavinod) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/receive/KafkaReceiver.java]: @ddingel thanks for your suggestions; updated to use union of common/local configs i.e effectiveLocalConfigs and addressed your comment

File: core/src/test/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelperTest.java
Comment (dandalavinod) [reply core/src/test/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelperTest.java]: @ddingel addressed your comment.

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (dandalavinod) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java]: @ddingel updated code to address your comments.

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (dandalavinod) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java]: @ddingel addressed your comment

**Comment by dandalavinod** [comment]
@ddingel - (1) i believe having the current number of runs with current default of 100L should be good enough and then if they need to increase the pollingTime - they could have the setting configured to their desired timeout (in properties instead of increasing the number of runs) which will be used once these code changes go in.
(2) Most likely  - user base might be using the the default setting of consumer.pollingTime that comes out of the properties files - in which case the default of 100L will be encountered not so commonly.

Hence i feel its less risk and more reliable for the consume operation to work without hitting the "kafka join in time" issue. 

Let me know, if anyone thinks that its imperative that we need to increase the number of runs ??

**Comment by dandalavinod** [comment]
@ddingel @nirmalchandra just a friendly reminder for your review, post my comments above

File: kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties
Comment (authorjapps) [inline kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties:33]: @dandalavinod , can you add an entry with `consumer.pollingTime` in this config file, which will override the default value of 500 and the test should pass.
e.g.
> consumer.pollingTime = 1000

Comment (ddingel) [reply kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties:33]: @authorjapps do I miss understand the current implementation, I thought the default now will be taken from KafkaConstants and will be 100ms?

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (ddingel) [inline core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:84]: Why do you need the ofNullable check here, if I understand it correctly getPollingTime already provides null safety and returns the default polling times as defined in KafkaConstants?

**Comment by ddingel** [comment]
@dandalavinod I think we should increase the value of runs to ensure stability over version upgrades.

As a user of some software I would find it very inconvenient to update to a new version to found out my test setup is broken because upstream decided to reduce some polling times.

Please increase the number of runs by 5 :).

File: kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties
Comment (dandalavinod) [reply kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties:33]: @ddingel i notice that effectiveLocalConfigs.getPollingTime() is returning NULL when its not set in properties. Sorry i should have tested this out early on without assuming. Hence you could see i explicitly set the default to 500L and believe would not require the number of runs increase you suggested.
Long durationForPolling = ofNullable(effectiveLocalConfigs.getPollingTime()).orElse(500L);

@authorjapps the properties file you referred is explicitly testing scenario when the property for pollingTime is not set; and i notice there is KafkaConsumeTest.java which refers to kafka_test_server.properties which has pollingTime set to 3000ms to cover the scenario where its explicitly set.

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (dandalavinod) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:84]: @ddingel i notice that effectiveLocalConfigs.getPollingTime() is returning NULL when its not set in properties. Sorry i should have tested this out early on without assuming. Hence you could see i explicitly set the default to 500L and believe would not require the number of runs increase you suggested.
Long durationForPolling = ofNullable(effectiveLocalConfigs.getPollingTime()).orElse(500L);

File: kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties
Comment (authorjapps) [reply kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties:33]: > the properties file you referred is explicitly testing scenario when the property for pollingTime is not set;

yep. 

>  which refers to kafka_test_server.properties which has pollingTime set to 3000ms 

understood.

Comment (authorjapps) [reply kafka-testing/src/test/resources/kafka_servers/kafka_test_server_polling.properties:33]: Hi @ddingel,
>@authorjapps do I miss understand the current implementation, I thought the default now will be taken from KafkaConstants and will be 100ms?

if e.g. `consumer.pollingTime = 1000` is provided, then it will use this(1000), otherise it defaults to `500 mili sec` which a constant hardcoded.

This change also means:
`consumer.pollingTime = 1000 or 100` if provided, will be used for both purpose.
1) initialPollWaitingForConsumerGroupJoin = 1000 or 100 milisec, retries 10 times
2) Subsequent pollingTime = 1000 or 100 milisec(configurable as usual)

Without this changes/earlier implementation:
1) initialPollWaitingForConsumerGroupJoin = 500 milisec HARD CODED, retries 10 times
   1.1)This is blocking other users like @dandalavinod as he needs slightly more time to rejoin e.g. 1000milisec
2) Subsequent pollingTime = configurable(100, 500, 1000, 5000 etc...)

Is this going to really affect you? If this is really a breaking change, we need to separate these two configs. But can you check n verify please?

**Comment by authorjapps** [comment]
> Please increase the number of runs by a factor 5 :). (from @ddingel )

@ddingel, Do you mean this will help in ConsumerGroupJoin?

@dandalavinod, can you please check if this helps!

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (authorjapps) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:84]: Hi @ddingel,
>@authorjapps do I miss understand the current implementation, I thought the default now will be taken from KafkaConstants and will be 100ms?

if e.g. `consumer.pollingTime = 1000` is provided, then it will use this(1000), otherise it defaults to `500 mili sec` which a constant hardcoded.

This change also means:
`consumer.pollingTime = 1000 or 100` if provided, will be used for both purpose.
1) initialPollWaitingForConsumerGroupJoin = 1000 or 100 milisec, retries 10 times
2) Subsequent pollingTime = 1000 or 100 milisec(configurable as usual)

> DEFAULT_POLLING_TIME_MILLI_SEC (as @sparrowV wrote )

@dandalavinod, can you use this please, this should help @ddingel and backward compatibility is retained upto some extent.
> org.jsmart.zerocode.core.kafka.KafkaConstants#DEFAULT_POLLING_TIME_MILLI_SEC


Without this changes/earlier implementation:
1) initialPollWaitingForConsumerGroupJoin = 500 milisec HARD CODED, retries 10 times
   1.1)This is blocking other users like @dandalavinod as he needs slightly more time to rejoin e.g. 1000milisec
2) Subsequent pollingTime = configurable(100, 500, 1000, 5000 etc...)

***

Comment (ddingel) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:84]: TLDR:
With getPollTime we use the correct default in all places and allow customisation. By increasing the number of runs we ensure that we do not change the default behaviour regarding kafka join, so that no user find themselves after doing an update with the new "Failed to join" messages in their tests.

---

Ahhh @dandalavinod, I think this solves the puzzle as @sparrowV pointed out the method to use is probably best:
```
    public static Long getPollTime(ConsumerLocalConfigs effectiveLocal) {
        return ofNullable(effectiveLocal.getPollingTime())
                .orElse(DEFAULT_POLLING_TIME_MILLI_SEC);
    }
```

This way we can ensure if at some point we change `DEFAULT_POLLING_TIME_MILLI_SEC` we get the initial polling as well :).


Hi @authorjapps, that the change will chain both constants and unify the handling to use a single value is a good thing. Users will be able to adjust. I was just confused where the `500 milisec` came from. As this is the previous hardcoded value :).

**Comment by ddingel** [comment]
@dandalavinod would it make sense to adjust the exception message as well, so that it maybe include a reference to the poll time out?

---

> > Please increase the number of runs by a factor 5 :). (from @ddingel )
> 
> @ddingel, Do you mean this will help in ConsumerGroupJoin?

Code as is does 10 tries a 500 secs, which works for us, if we now go down to 10 tries a 100 sec this might be too little. 
Which would at least force me to adjust that in my settings. And there might be other users like me with a similar latency. At least currently we bail out with an exception, indicating into the right direction.

But in the end this is all some sort of band-aid, we constantly join and leave the consumer group and the only way to check if this was successful is checking if we already have some assignment... Maybe we need to rethink this?

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (dandalavinod) [reply core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java]: Done - updated code to reuse getPollTime function

**Comment by dandalavinod** [comment]
@ddingel @sparrowV @authorjapps made suggested updates and tweaked the exception message a bit let me know if any suggestions on better wording.

**Comment by dandalavinod** [comment]
> @dandalavinod, can you please check if this helps!

i believe it will add more resiliency on top of the pollingTime fix being delivered here; hence changed the runs to 50 to be safe in rare case where default of 100ms polling timeout is picked up.

**Comment by ddingel** [comment]
@dandalavinod thanks for all the changes. For me this code looks fine. Not sure if a test is missing or not. TBH I have a hard time following all this threading/non threading conversations.

I would say lets wait for @authorjapps or @sparrowV final approval :).

**Comment by sparrowV** [comment]
looks good to me,  If @authorjapps  does not have any other change requests, I think its ready to be merged.

**Comment by authorjapps** [comment]
Thanks, @dandalavinod for taking care of overall concerns and @ddingel's advices in terms of backward compatibility. Thanks @sparrowV for your timely suggestions!

@sparrowV @ddingel , Can you please approve it now?
CI link(travis): https://travis-ci.org/github/authorjapps/zerocode/builds/747266282

CI is green now. Happy to mergee it soon after approvals.

**Review by authorjapps — APPROVED** [review]

File: core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java
Comment (authorjapps) [inline core/src/main/java/org/jsmart/zerocode/core/kafka/helper/KafkaConsumerHelper.java:94]: 👍 🏆

**Review by sparrowV — APPROVED** [review]

**Review by ddingel — APPROVED** [review]

**Review by nirmalchandra — APPROVED** [review]

**Pull request merged**
